helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -497,6 +499,67 @@ public boolean isInMaintenanceMode(String clusterName) {
         .exists(keyBuilder.maintenance().getPath(), AccessOption.PERSISTENT);
   }
 
+  @Override
+  public void enableClusterPauseMode(String clusterName, boolean cancelPendingST, String reason) {
+    logger.info("Enable cluster pause mode for cluster: {}. Cancel pending ST: {}. Reason: {}.",
+        clusterName, cancelPendingST, reason);
+
+    PauseSignal pauseSignal = new PauseSignal("pause");
+    pauseSignal.setPauseCluster(Boolean.toString(true));
+    pauseSignal.setCancelPendingST(cancelPendingST);
+    pauseSignal.setFromHost(NetworkUtil.getLocalhostName());
+    pauseSignal.setTriggerTime(Instant.now().toString());
+    if (reason != null && !reason.isEmpty()) {
+      pauseSignal.setReason(reason);
+    }
+
+    BaseDataAccessor<ZNRecord> baseDataAccessor = new ZkBaseDataAccessor<>(_zkClient);
+    HelixDataAccessor accessor = new ZKHelixDataAccessor(clusterName, baseDataAccessor);
+
+    if (baseDataAccessor.exists(accessor.keyBuilder().pause().getPath(), AccessOption.PERSISTENT)) {
+      throw new HelixPauseExitsException(clusterName + " pause signal already exits");
```

Comment:
1. The description does not necessarily match the Exception class name. Do you mean "exits" or "exists"?
2. If pause mode has existed, shall we consider the operation is done? But I agree that we should not just return nothing, since it might be paused due to a different reason.

This makes me think about whether or not we shall enter pause mode if the cluster is in controller pause or maintenance mode. My preference is don't mix things up. Reject the call under any confusing situations. So you have to check for the other signals here.

Reply:
Ah typo... Should be Exists.
Right, if it already exists, because we have different modes for the PAUSE znode, we don't want to overwrite it. So the PauseExistsException tells the znode already exists and rejects the request as a clear way.

Reply:
So do you agree that we should fail the operation even if the cluster is only in the Maintenance mode?